### ISSUE TITLE
--no-recreate flag shows "invalid container format"

### ISSUE THREAD

**Issue by KuanYuLai** [docker/compose#8940]
**Description**

I want to scale my docker container without recreating the existing one. As a result, I add the ```--no-recreate``` flag in my docker-compose command. 
```
docker-compose up -d --scale api=2 --no-recreate
``` 
However, I received this error while creating the container 
```
Error response from daemon: Invalid network mode: invalid container format container:<name|id>
```
This problem seems to happen after upgrade my docker-compose from  1.28.0 to  2.1.1.

Is it related to the naming scheme change from ```<project>_<service>_<index>``` to ```<project>-<service>-<index>``` ?

**Steps to reproduce the issue:**
1. Use a docker-compose file to start a service first
2. Scale the service with --scale flag and --no-recreate flag

**Describe the results you received:**
Error response from daemon: Invalid network mode: invalid container format container:<name|id>

**Describe the results you expected:**
docker create additional container without recreate the existing noe

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker compose version`:**

```
Docker Compose version v2.1.1
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 11
  Running: 11
  Paused: 0
  Stopped: 0
 Images: 9
 Server Version: 20.10.7
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d71fcd7d8303cbf684402823e425e9dd2e99285d
 runc version: b9ee9c6314599f1b4a7f497e1f1f856fe433d3b7
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.4.0-81-generic
 Operating System: Ubuntu 20.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.775GiB
 Name: llyServer
 ID: KTEW:WVJD:SCPK:X2VQ:LGVA:ML2B:WXAU:QHFT:XOU6:DSJE:MGXM:IOHS
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: true
 Default Address Pools:
   Base: 192.168.0.0/16, Size: 24

WARNING: No swap limit support
```

**Additional environment details:**
I specify the project name in the .env file
```
PROJECT=myproject
```

**Comment by KuanYuLai** [comment]
Update: This issue seems to occur after v2.0.1

**Comment by nocive** [comment]
@KuanYuLai how about in the latest pre release (ie: 2.2.0) ?

**Comment by KuanYuLai** [comment]
@nocive The pre release v2.2.0 is still having this issue, but v2.2.1 seems working now.

**Comment by navels** [comment]
I'm getting this error with v2.2.1.

**Comment by Alhel** [comment]
in 2.2.2 the same error

**Comment by HarrySky** [comment]
This error appears after 2.1.0, i can `docker compose -p "project-name" -f "${compose_file}" up -d --no-recreate` on docker compose v2.0.1

**Comment by HarrySky** [comment]
To reproduce you need this minimal `docker-compose.yml` file:
```
version: "3.9"
services:
  service-1:
    image: alpine:3.15.0
    container_name: service-1-local
    hostname: service-1-local
    restart: always
    networks:
      services-local:
        ipv4_address: 172.25.30.101
    command: "sleep 120"

  service-2:
    image: alpine:3.15.0
    depends_on:
      - service-1
    container_name: service-2-local
    hostname: service-2-local
    restart: always
    networks:
      services-local:
        ipv4_address: 172.25.30.102
    command: "sleep 120"

networks:
  services-local:
    name: services-local
    driver: bridge
    ipam:
      driver: default
      config:
        - subnet: "172.25.30.0/24"
```

1. Bring up first service via `docker compose -p "project-name" -f "docker-compose.yml" up -d service-1`
2. Fail to bring other services via `docker compose -p "project-name" -f "docker-compose.yml" up -d --no-recreate`

You will get `Error response from daemon: Invalid network mode: invalid container format container:<name|id>` after second step

**Comment by HarrySky** [comment]
Maybe this PR is related to the bug https://github.com/docker/compose/pull/8792 , as it seems to be the only one that changed networking code between 2.0.1 -> 2.1.0. But I can't be sure.

Could somebody please look at this ( @mat007 ) ? Reproduction steps are above. Can't use Docker Compose above 2.0.1 because of this bug.

Maybe there are some symbols in network/service/project names we should avoid?

**Comment by oiljin** [comment]
in v2.2.3 the same error

**Comment by aminvakil** [comment]
I have faced this in https://github.com/getsentry/self-hosted/pull/1393 too.

1.28.0: https://github.com/getsentry/self-hosted/runs/5633041888?check_suite_focus=true#step:4:1936
1.29.2: https://github.com/getsentry/self-hosted/runs/5633041941?check_suite_focus=true#step:4:1928
2.2.3: https://github.com/getsentry/self-hosted/runs/5633042067?check_suite_focus=true#step:4:1946

FYI it worked on 2.0.1: https://github.com/getsentry/self-hosted/runs/5633041984?check_suite_focus=true

**Comment by sophos-rickc** [comment]
I am experiencing this in Docker Compose 2.3.3

**Comment by aminvakil** [comment]
It seems that it has been fixed in 2.3.4 (https://github.com/getsentry/self-hosted/runs/5708677698?check_suite_focus=true#step:4:2177).

**Comment by HarrySky** [comment]
Problem can no longer be reproduced in `v2.4.1` via instructions I provided above :rocket:
